### ISSUE TITLE
Comparison of JP2 and other data holdings at different locations

### ISSUE THREAD

**Issue by wafels** [Helioviewer-Project/api#340]
There are four different JPIP servers running - GSFC, ROB, IAS and ESAC. Each of them are serving the data stored at that location. It would be good to understand more completely which data are at which location. This would improve the ability of each helioviewer location to fill in gaps in their data holdings.

**Comment by dgarciabriseno** [comment]
The leading idea for this is from Bogdan's message [here](https://github.com/Helioviewer-Project/api/issues/325#issuecomment-1716037046)

> My idea is to have insertion time as a column in the HV database and add the ability to query that remotely (HAPI?)

So the idea is something like this:
![diagram of mirror algorithm](https://www.plantuml.com/plantuml/png/PL4nJiGm4EpzYj5LIGdbNS4L22eAdc2I2R7aU2rxZ-Y4-3ih9o0TDCdEh3qxipvNzKKlItIjSCVN2yiLkOJ5sx-oVB1WaeB45LE8h2UNxMJG80c79-lG3tDh7V3f2ZEzwJnJ6uUHwaEqMs525-f6cjX33vsPq5rPUpWW-ZT6rEm7tjt2-ryuofgY-D54QEJnSSTDzUGOAz4bwQstZzEmqA8j6OiKed1WqflP3lYHJ-JOgiQzYyhVYKnZcDps_MzfJuzrDe3kHGgtdK45kSZ0QbdDrHOwYMvklspxajtPTEqflW40)

If insertion time is the best metric to use, there are some things to think about:
- Only works one-way. When B mirrors from A, the timestamps in B will be newer then A.
- I'm not sure if there could be issues if working in both mirror mode + pulling from upstream sources

@bogdanni @ebuchlin, any thoughts?

**Comment by ebuchlin** [comment]
Hello, thanks for the proposal! We definitely need a better synchronization process.

A few clarifications first about your flowchart:

* Is "query primary server for new files" about ingestion date being more recent than the last (remote) ingestion date of the files already received (by the local server)? (I guess yes)
* In the different servers' databases, is ingestion date the ingestion date on the local server, or first ingestion date of this version of the file on any server?
* In the box "is timestamp more recent?", is this about the remote file timestamp being more recent than the local file file timestamp? (I guess yes) But then how is the remote file timestamp obtained by the local server?
* What is happening to the local database after download, do we keep the same file ingestion mechanism as currently?

A proposal:

* The "ingestion date" (everything in UTC) in the local database is the local ingestion date, which should be more or less equivalent to the local file timestamp (and same for the remote server, which is "local" from its own viewpoint)
* The query to the remote server is by ingestion date range (usually, in routine operations: from the last locally processed remote ingestion date to now)
  * This query uses the remote server's HAPI interface
  * There could be other conditions in the query (observation date range, dataset...) because maybe we don't want to update everything at once, but then the last locally processed remote ingestion date cannot be stored to be used later anymore.
* For each file in the result of this query, ordered by ingestion date, if (the file does not exist in the local database) or (the local ingestion date is older than the remote ingestion date):
  * download file
  * ingest it (setting the ingestion date to the current date; replacing existing database row if there is one)
* There could be a mechanism to check that the servers' times (clocks) are synchronized as expected

**Comment by dgarciabriseno** [comment]
**Clarifications**:

> Is "query primary server for new files" about ingestion date being more recent than the last (remote) ingestion date of the files already received (by the local server)? (I guess yes)

Yes, new files means the ingestion date on the remote server is newer than the ingestion date on the local server. Since after the local server pulls the new file, its local ingestion date will be newer than the remote ingestion date, this method only works one way. So there must be a "primary" or "authoritative" server.

> In the different servers' databases, is ingestion date the ingestion date on the local server, or first ingestion date of this version of the file on any server?

It is the ingestion date of the file on the local server.

> In the box "is timestamp more recent?", is this about the remote file timestamp being more recent than the local file file timestamp? (I guess yes) But then how is the remote file timestamp obtained by the local server?

Yes, it will be one of the parameters returned in query.

> What is happening to the local database after download, do we keep the same file ingestion mechanism as currently?

Yes, the existing ingestion method would stay the same. The part this proposal changes is how new files are selected. For this download method, instead of querying a web directory, it would query another helioviewer server and filter the results following the proposal we're discussing here.

**Proposal Comments**:
> The "ingestion date" (everything in UTC) in the local database is the local ingestion date, which should be more or less equivalent to the local file timestamp (and same for the remote server, which is "local" from its own viewpoint)

Agreed, UTC for all dates. And the date of interest is the time the file was added to the local database.

> There could be other conditions in the query (observation date range, dataset...) because maybe we don't want to update everything at once, but then the last locally processed remote ingestion date cannot be stored to be used later anymore.

It sounds like the dates will need to be stored per-source. Since the database is already storing the latest ingestion dates, it could get the latest ingestion dates for the sources being updated, and choose the oldest time from that selection.

Updated diagram:
![Updated diagram for sync method](http://www.plantuml.com/plantuml/png/RL51JiCm4Bpx5QEUqaCFzDXAq0D2AYeILF00DvaqHeutibsY2l5tjab1KFMIxEpQC-FPLSesvaFhpL2OnSk1yOWkkjRgFp5-CAAMYC0UjVDCIpExnPRv44E2Tob3wc42sRspp0czourIICUO7TF2O8R7vpdUjXijzMuhcLY7FLDs4l3gMfgbwQmU7ngJ1nMaBPk1Tu9F4zdHPkDg5EiqgWNsTrFaXW75aMaA0tYxeqVgRCcR_t3-2olfWnTRdKYqkRW_uu5rQUWJKGIPwcoKksxufC-e-ZPW8wNr5_E_rWpmenx8QgY6tbwoN1fYg5njnkyLnsCIw1i5A9uaSbmBifmoBvaq2fVEcGJ9usE-TC7IcPNowhQ_0G00)

**Comment by bogdanni** [comment]
I'm away and I could not think about this schema.

Another thing to add is the computation of a checksum for each file ingested. This checksum to be stored as a column and can be retrieved over HAPI. This allows the following data integrity checks:
- identify local storage corruption
- identify when a file on a remote server is different

**Comment by ebuchlin** [comment]
So the query could select remote files based on their (remote) ingestion time, but the comparison between remote files and local files could be on checksum only? (no need to compare ingestion times if the checksums are compared)

**Comment by dgarciabriseno** [comment]
Checksum sounds good. That seems more reliable than just checking a timestamp.

**Comment by dgarciabriseno** [comment]
I wrote up what we've discussed so far [here](https://github.com/Helioviewer-Project/api/wiki/Mirroring). And I'm sure I've made some assumptions particularly about who's mirroring which sources.

Please review and feel free to edit.

For the HAPI server, how will the datasets be grouped? We could group by source id, then each HAPI dataset would be at the measurement level i.e. AIA 94 is its own dataset, AIA 304 is its own dataset, etc

**Comment by ebuchlin** [comment]
Looks good. A small comment (somewhat following [this comment](https://github.com/Helioviewer-Project/api/issues/340#issuecomment-1790811255)): the query on the primary server needs to select by ingestion time, but I don't see why returning ingestion time is necessary. In pseudo SQL: `SELECT name, ingestion_time, jp2_url WHERE ingestion_time > ...`, no need for `SELECT ingestion_time, name, ...`. 

I don't have an opinion on dataset grouping.

**Comment by dgarciabriseno** [comment]
Makes sense. Technically in HAPI there's no way to turn that off, though. Time is always returned even if it's not requested.
